// ecb-currency-converter/dist/currencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * An enum with all supported currencies.
 * @type {{EUR: string; USD: string; JPY: string; BGN: string; CZK: string; DKK: string; GBP: string; HUF: string; PLN: string; RON: string; SEK: string; CHF: string; ISK: string; NOK: string; HRK: string; RUB: string; TRY: string; AUD: string; BRL: string; CAD: string; CNY: string; HKD: string; IDR: string; ILS: string; INR: string; KRW: string; MXN: string; MYR: string; NZD: string; PHP: string; SGD: string; THB: string; ZAR: string}}
 */
exports.Currencies = {
    EUR: 'EUR',
    USD: 'USD',
    JPY: 'JPY',
    BGN: 'BGN',
    CZK: 'CZK',
    DKK: 'DKK',
    GBP: 'GBP',
    HUF: 'HUF',
    PLN: 'PLN',
    RON: 'RON',
    SEK: 'SEK',
    CHF: 'CHF',
    ISK: 'ISK',
    NOK: 'NOK',
    HRK: 'HRK',
    RUB: 'RUB',
    TRY: 'TRY',
    AUD: 'AUD',
    BRL: 'BRL',
    CAD: 'CAD',
    CNY: 'CNY',
    HKD: 'HKD',
    IDR: 'IDR',
    ILS: 'ILS',
    INR: 'INR',
    KRW: 'KRW',
    MXN: 'MXN',
    MYR: 'MYR',
    NZD: 'NZD',
    PHP: 'PHP',
    SGD: 'SGD',
    THB: 'THB',
    ZAR: 'ZAR'
};
